File: 0021_merge_two_sorted_linked_lists.kt
// https://neetcode.io/problems/merge-two-sorted-linked-lists

// Iterative solution. Constant space complexity and O(n+m) time complexity, where n & m are the sizes of the input lists.
/**
 * Example:
 * var li = ListNode(5)
 * var v = li.`val`
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */
class Solution {
    fun mergeTwoLists(list1: ListNode?, list2: ListNode?): ListNode? {
        val dummyHead = ListNode(Int.MIN_VALUE)
        var tail: ListNode? = dummyHead

        var pointer1 = list1
        var pointer2 = list2

        while (pointer1 != null && pointer2 != null) {
            if (pointer1.`val` < pointer2.`val`) {
                tail?.next = pointer1
                pointer1 = pointer1.next
            } else {
                tail?.next = pointer2
                pointer2 = pointer2.next
            }

            tail = tail?.next
        }

        tail?.next = pointer1 ?: pointer2
        return dummyHead.next
    }
}

File: 0021_merge_two_sorted_linked_lists.cpp
// https://leetcode.com/problems/merge-two-sorted-lists/

// O(n+m) time complexity with a constant space complexity. Iterative solution.
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
        ListNode* dummyNode = new ListNode(INT_MIN);
        ListNode* tail = dummyNode;

        ListNode* ptr1 = list1;
        ListNode* ptr2 = list2;

        while (ptr1 && ptr2) {
            if (ptr1->val < ptr2->val) {
                tail->next = ptr1;
                ptr1 = ptr1->next;
            } else {
                tail->next = ptr2;
                ptr2 = ptr2->next;
            }

            tail = tail->next;
        }

        tail->next = ptr1 ? ptr1 : ptr2;

        return dummyNode->next;
    }
};
